tools/codegen/gen_functionalization_type.py

```diff
@@ -0,0 +1,426 @@
+    else:
+        forward_lambda_call_str = f'at::{forward_name}({", ".join(forward_lambda_exprs)}){maybe_idx}'
+
+    view_meta_str = f"""
+      at::functionalization::ViewMeta view_meta = at::functionalization::ViewMeta(
+        [{captured_view_args_str}]({', '.join(a.decl() for a in forward_lambda_args)}) -> {forward_ret_type.cpp_type()} {{
+          return {forward_lambda_call_str};
+        }},
+        [{captured_view_args_str}]({', '.join(a.decl() for a in reverse_lambda_args)}) -> {reverse_ret_type.cpp_type()} {{
+          return at::functionalization::impl::{reverse_name}({", ".join(reverse_lambda_exprs)});
+        }}
+      );"""
+
+    api_name = CppSignatureGroup.from_native_function(
+        f, method=False, fallback_binding=f.manual_cpp_binding).most_faithful_signature().name()
+    if f.tag is not None and f.tag is Tag.inplace_view:
+        # This op is both an inplace op AND a view op.
+        # See Note [Functionalization Pass - Inplace View Ops] for details.
+        # I currently have the view meta call into the out-of-place variant of the view, to avoid
+        # having to define an extra ~20 inplace {view}_inverse_ functions.
+        # Most view ops don't have NativeFunctionGroup's both, because we don't define out= variants for view ops.
+        # I'm assuming that every inplace-view op has a corresponding out-of-place view op,
+        # with the same name but the trailing underscore removed.
+        # This is currently asserted at parse time in gen.py (see error_check_native_functions).
+        assert f.func.kind() is SchemaKind.inplace
+        api_name = api_name[:-1]
+
+    # The meta API call should use the same arguments, but convert all tensors to meta tensors first.
+    meta_conversion_str, meta_call_ctx = convert_to_meta_tensors(dispatcher_sig)
+    meta_call_args = [e.expr for e in translate(meta_call_ctx, dispatcher_sig.arguments(), method=False)]
+    if Variant.method in f.variants:
+        assert is_tensor_like(meta_call_ctx[0].argument)
+        meta_call_str = f'{meta_call_ctx[0].name}.{api_name}({", ".join(meta_call_args[1:])})'
+    else:
+        meta_call_str = f'at::{api_name}({", ".join(meta_call_args)})'
+
+
+    if f.tag is not None and f.tag is Tag.inplace_view:
+        # See Note [Functionalization Pass - Inplace View Ops] for more details
+        return f"""
+      {view_meta_str}
+      at::functionalization::impl::mutate_view_meta(self, view_meta);
+      {unwrap_tensor_args_str}
+      {meta_conversion_str}
+      {return_type} reference_tensor_output;
+      {{
+        at::AutoDispatchSkipFunctionalize guard;
+        reference_tensor_output = {meta_call_str};
+      }}
+      // See  Note [Propagating strides in the functionalization pass]
+      at::functionalization::impl::set_strides(self, reference_tensor_output);
+      return self;
+"""
+
+    else:
+        return f"""
+      {unwrap_tensor_args_str}
+      {meta_conversion_str}
```

Comment:
There isn't any reason for meta conversion to happen here; you can do it inside the limited block scope for computing reference tensor, right? That helps reduce the scope of the meta bindings and means you can do the codegen for this portion of the logic in its own helper function, reducing the size of this function.

Reply:
The location of the meta conversion is also confusion, because ordinarily I would expect if I do "do some bindings; do some more bindings", I expect the code doing the generation and binding to proceed in that order. But here all of the work for building view_meta_str is done first, and then the meta construction is done as an afterthought (probably, because it was done after!)